### PR TITLE
feat (core/cs): Make `.go xyz` ignore non-numeric data to improve copy-paste usability

### PR DESCRIPTION
**Pull request by kjack9**
## Changes Proposed:

This PR proposes changes to:
-  [X] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

Rewrites the command handler function for `.go xyz` to handle flexible inputs. The new implementation will use regex to extract float and integer values from the provided arguments and will discard any other characters.

Since the parsing doesn't care about any of the surrounding text, many different versions of coordinates are supported by this implementation. The only requirement is that the numbers are in the right order (`x y [z [mapId [orientation]]]`) and have at least one delimiter character (not a number) between each number.

The following all work as expected:
`.go xyz 4073.1 -3536.1 123.5 329 2.8` - previously-required format
`.go xyz 4073.2-3536.2,123.5` - any non-number works as a delimiter
`.go xyz Position: X: 4073.3 Y: -3536.3 Z: 123.5` - sniff parse format
`.go xyz Position: X: 4073.4 Y: -3536.4 Z: 123.5 329 O: 2.8` - sniff parse format with `mapId` inserted
`.go xyz (4073.5f,-3536.5f,123.5f` - C++ format, doesn't have to be complete
`.go xyz Position(4073.6f, -3536.6f, 123.5f, 329 2.8f)` - C++ position with `mapId` inserted
`.go xyz 4073.7-3536.7f123.5?329` - any non-number works as a delimiter
`.go xyz This IsAStupidTextBut4073.8ItWorks Fine-3536.8BecauseOf123.5RegExParsing329`

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [X] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

`.tel strath` -- puts you in the correct map for the commands that don't specify one

Try each of these cases one at a time. Each one should move you diagonally slightly so you can tell it worked. Some will have orientation and others won't.

`.go xyz 4073.1 -3536.1 123.5 329 0.1`
`.go xyz 4073.2-3536.2,123.5`
`.go xyz Position: X: 4073.3 Y: -3536.3 Z: 123.5`
`.go xyz Position: X: 4073.4 Y: -3536.4 Z: 123.5 329 O: 0.1`
`.go xyz (4073.5f,-3536.5f,123.5f`
`.go xyz Position(4073.6f, -3536.6f, 123.5f, 329 0.1f)`
`.go xyz 4073.7-3536.7f123.5?329`
`.go xyz This IsAStupidTextBut4073.8ItWorks Fine-3536.8BecauseOf123.5RegExParsing329`

Feel free to bring in your own test cases. As long as the numbers are in the correct order (`x, y, [z [mapId [orientation]]]`) with at least one non-number character between each number it should work fine.

## Known Issues and TODO List:

The ordering of parameters is weird when copy-pasting. I'd like to find a way to make the mapId be "whatever the last argument is" while still giving the option to omit it entirely. I considered making is so that `x y z o` all need to be in float format (`123.0` rather than `123`), so then I can just find the last number that isn't in that format, but I want feedback before I implement that.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by sudlud** [comment]
Looking forward to this 🙃

File: src/server/scripts/Commands/cs_go.cpp [hunk 1]
```diff
@@ -288,34 +291,68 @@ class go_commandscript : public CommandScript
...
-    static bool HandleGoXYZCommand(ChatHandler* handler, float x, float y, Optional<float> z, Optional<uint32> id, Optional<float> o)
+    /**
+     * @brief Teleports the GM to the specified world coordinates, optionally specifying map ID and orientation.
+     *
+     * @param handler The ChatHandler that is handling the command.
+     * @param inputCoords The coordinates to teleport to in format "x y z [mapId [orientation]]".
+     * @return true The command was successful.
+     * @return false The command was unsuccessful (show error or syntax)
+     */
+    static bool HandleGoXYZCommand(ChatHandler* handler, Tail inputCoords)
```
Comment (Kitzunu) [inline src/server/scripts/Commands/cs_go.cpp:302]:
```suggestion
    static bool HandleGoXYZCommand(ChatHandler* handler, Tail args)
```

File: src/server/scripts/Commands/cs_go.cpp [hunk 2]
```diff
@@ -288,34 +291,68 @@ class go_commandscript : public CommandScript
...
-            if (!MapMgr::IsValidMapCoord(mapId, x, y))
-            {
-                handler->PSendSysMessage(LANG_INVALID_TARGET_COORD, x, y, mapId);
-                handler->SetSentErrorMessage(true);
-                return false;
-            }
-            Map const* map = sMapMgr->CreateBaseMap(mapId);
-            z = std::max(map->GetHeight(x, y, MAX_HEIGHT), map->GetWaterLevel(x, y));
+            return false;
+        }
```
Comment (Kitzunu) [inline src/server/scripts/Commands/cs_go.cpp:335]: There should be an error message if not x and y is provided.

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/scripts/Commands/cs_go.cpp (hunk 2 above)
Comment (kjack9) [reply src/server/scripts/Commands/cs_go.cpp:335]: The current solution causes the syntax help to be shown. What error would you like me to show instead?

File: src/server/scripts/Commands/cs_go.cpp (hunk 1 above)
Comment (kjack9) [reply src/server/scripts/Commands/cs_go.cpp:302]: Fair enough, I'll change this.

**Comment by kjack9** [comment]
Wiki update: https://github.com/azerothcore/wiki/pull/935

File: src/server/scripts/Commands/cs_go.cpp (hunk 2 above)
Comment (Kitzunu) [reply src/server/scripts/Commands/cs_go.cpp:335]: handler->PSendSysMessage(LANG_INVALID_TARGET_COORD, x, y, mapId);
            handler->SetSentErrorMessage(true);
            return false;

Comment (kjack9) [reply src/server/scripts/Commands/cs_go.cpp:335]: I'm so sorry, I don't understand.

At the point in the function that you're suggesting this, I don't have `x`, `y`, or `mapId` yet. In fact, we know that the user hasn't provided at least 2 values, so `x` or `y` are going to be null. Emitting `LANG_INVALID_TARGET_COORD` here also means that it's impossible to get access to the syntax help.

I _do_ check that the coordinates provided are valid further down, and if not I send the `LANG_INVALID_TARGET_COORD` message with the `x`, `y`, and `mapId` values I was able to parse out.
https://github.com/kjack9/azerothcore-wotlk/blob/ca1435384192fc05331b32da82dea69be486d755/src/server/scripts/Commands/cs_go.cpp#L348-L353

Sorry to be missing this one. Let me know what you're wanting in this situation.

Comment (heyitsbench) [reply src/server/scripts/Commands/cs_go.cpp:335]: Most existing commands in my experience do stick the syntax in your face as an error message when improper syntax is supplied, so personally at least I agree with kjack's current implementation.

**Comment by kjack9** [comment]
It's not a blocker to this PR merging, but I'd be interested in comments on this potential todo item:

> The ordering of parameters is weird when copy-pasting. I'd like to find a way to make the mapId be "whatever the last argument is" while still giving the option to omit it entirely. I considered making is so that x y z o all need to be in float format (123.0 rather than 123), so then I can just find the last number that isn't in that format, but I want feedback before I implement that.

File: src/server/scripts/Commands/cs_go.cpp (hunk 2 above)
Comment (Kitzunu) [reply src/server/scripts/Commands/cs_go.cpp:335]: this would give the syntax too. And keep the same behavior as before.

Comment (Kitzunu) [reply src/server/scripts/Commands/cs_go.cpp:335]: never mind my brain is not functioning correctly

**Review by Kitzunu — APPROVED** [review]

**Comment by Kitzunu** [comment]
> x y z o all need to be in float format

No, this will just make the command bad to use.

**Comment by kjack9** [comment]
Thanks to all the reviewers! Anything else on this PR before it's good to go?

**Pull request merged**